h/schemas/base.py

```diff
@@ -99,6 +100,61 @@ def serialize(self, node, appstruct):
     return EnumType
 
 
+def _colander_exception_msg(exc):
+    """
+    Combine error messages from a `colander.Invalid` exception.
+
+    :type exc: colander.Invalid
+    :rtype str:
+    """
+    msg_dict = exc.asdict()
+    for child in exc.children:
+        msg_dict.update(child.asdict())
+    msg_list = ["{}: {}".format(field, err) for field, err in msg_dict.items()]
+    return "\n".join(msg_list)
+
+
+def validate_query_params(schema, params):
+    """
```

Comment:
I got this docstring to render with:

```
Validate query parameters using a Colander schema.

Repeated fields in the input are either preserved or discarded except for
the last value depending on whether the corresponding schema node is
a sequence.

:type schema: :py:class:`colander.Schema`
:param params: Query parameter dict, usually ``request.params``
:type params: :py:class:`webob.multidict.MultiDict`
:rtype: :py:class:`webob.multidict.MultiDict`
:raises: :py:class:`ValidationError`
```

Reply:
Looking at the output of `make docstrings`, the `:py:class:` and quotes around the type names don't appear to make a difference to the rendered output. For example Sphinx links to the `ValidationError` type and renders the third-party types the same way with or without these. This is what I'd expect since the fact that these are references to types is something I'd hope Sphinx should be able to infer from the context.

If it isn't necessary to use `:py:class:`-style annotations I would find the code more readable if we could leave them out.

Reply:
That particular bit is totally up to you (this stuff is rather complicated)...feel free to simplify if you'd like (but use double-ticks for code in the body of the docstring).

There were a few things in your original docstring that broke rendering, however. @seanh is really the expert in this; I'm still learning. My main objective is to get stuff to render right...

Reply:
Oh wait. Ignore the above. I got confused by the excruciatingly slow live reload process of our current `make docstrings` task.

Reply:
To summarize the Slack discussion, adding `:py:class:` everywhere would add a lot of noise. and make it harder to read the code. Sphinx renders `:type foo: h.foo.Bar` differently than ``:type foo: :py:class:`h.foo.Bar` `` but the cross-links still work with the former. On that basis I'm going to suggest sticking with the simpler form.